deep_qa/data/vocabulary.py

```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+import codecs
+import logging
+
+import tqdm
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class Vocabulary:
+    """
+    A Vocabulary maps strings to integers, allowing for strings to be mapped to an
+    out-of-vocabulary token.
+
+    Vocabularies are fit to a particular dataset, which we use to decide which tokens are
+    in-vocabulary.
+
+    Vocabularies also allow for several different namespaces, so you can have separate indices for
+    'a' as a word, and 'a' as a character, for instance.  Most of the methods on this class allow
+    you to pass in a namespace; by default we use the 'tokens' namespace, and you can omit the
+    namespace argument everywhere and just use the default.
+    """
+    def __init__(self):
+        self._padding_token = "@@PADDING@@"
+        self._oov_token = "@@UNKOWN@@"
+        self._token_to_index = defaultdict(lambda: {self._padding_token: 0, self._oov_token: 1})
+        self._index_to_token = defaultdict(lambda: {0: self._padding_token, 1: self._oov_token})
+        self._finalized = False
+
+    def set_from_file(self, filename: str, oov_token: str="@@UNKNOWN@@", namespace: str="tokens"):
```

Comment:
Sounds good, and agree on the usefulness of adding some additional words to a loaded model.  I wasn't sure if you were anticipating supporting that or not.